### PR TITLE
Flush the session in `handle_failure()` prior to invoking callbacks

### PR DESCRIPTION
**Pull request by sydykov**
Related: #27615. The full description can be found there.

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -1808,13 +1808,13 @@ def handle_failure(
...
-        if callback and context:
-            self._run_finished_callback(callback, context, callback_type)
-
         if not test_mode:
             session.merge(self)
             session.flush()
 
+        if callback and context:
+            self._run_finished_callback(callback, context, callback_type)
+
```
Comment (ephraimbuddy) [inline airflow/models/taskinstance.py:1817]: Can you add test? Looks like something that can be tested

Comment (sydykov) [reply airflow/models/taskinstance.py:1817]: I can try; that's all I can promise 🙂

Comment (ephraimbuddy) [reply airflow/models/taskinstance.py:1817]: @sydykov, any update on this?

Comment (sydykov) [reply airflow/models/taskinstance.py:1817]: @ephraimbuddy, hi! I am planning to get back to this in January. Thank you for asking, and I regret delaying the PR this much.

**Comment by potiuk** [comment]
conflicts need solving. Also it would be good to add description in the commit message. The issue in GitHub might or might not exist in the future, and we should add contex to a change in the commit message.

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (Taragolis) [inline airflow/models/taskinstance.py:1817]: I don't think this is the right way for solve #27616.

If we call `sesion.flush` before callbacks we will lock row in DB until callback completed as result you won't able to reset task from the UI as well as other side effects related to this issue.

How to reproduce
```python
import time

from airflow.decorators import task
from airflow.models import DAG
import pendulum


def endless_callback(context):
    while True:
        print("Kill me if you can!")
        time.sleep(10)


with DAG(
    dag_id="infinity_callback",
    schedule=None,
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    catchup=False,
    tags=["infinity callback", "lock db"],
):
    @task(on_retry_callback=endless_callback, retries=1)
    def failed_task():
        1 / 0

    x = failed_task()
```

Ideally we should commit changes before run any callbacks but we can't do this - `LocalTaskJob` would decide that the state is set externally and kill process with callback.

The general issues with callbacks right now if callback run for a long time then:
1. DB Session doesn't return into the poll
2. Our transaction would be idled which not a good (I'd rather say terrible) and lead to a lot different problems.

I'm not sure there is good solution exist without decouple "callbacks" from Task Runner

**Review by Taragolis — CHANGES_REQUESTED** [review]

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (sydykov) [reply airflow/models/taskinstance.py:1817]: @ephraimbuddy, hello. I am still traveling and haven't had a chance to get back to this yet, unfortunately.

Comment (sydykov) [reply airflow/models/taskinstance.py:1817]: This is an interesting point. I didn't consider long-running callbacks... The proposed little change can serve as a temporary solution for short, quick callbacks (at least) that need to affect the state. What do you think?

Comment (Taragolis) [reply airflow/models/taskinstance.py:1817]: Actually this is the problem with current implementation callbacks as well as email notification. 
It run inside of task instance job and depend on it.

Let me explain some options.

Option 1. Make changes and commit them before run callbacks. On next heartbeat LocalTaskJob decided that status changed externally and terminate process.

Option 2. Make changes but do not commit changes until callbacks finished. We lock TI record for unknown time. And in Airflow 2.6.0 we add ability to run multiple callbacks one after each other as well as some providers grant ability to send Notifications by callbacks

Option 3. Keep it as it now. No Task Instance status updates. DB also will keep in transaction until callback(s) finished. Still not good, IMHO it is less harmful for now.

Comment (Taragolis) [reply airflow/models/taskinstance.py:1817]: Anyway, if other decided than Option 2 or Option 1 better than current Option 3 than I would dismiss change request.

IMHO, we can't do anything better:
- without refactor TI + LocalTaskJob
- or decouple callback outside of Task Execution
- or even transform all callbacks (and email notifiers) to some simple version of TI and execute as regular tasks in worker

**Pull request closed without merging**